src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
     _shard_quota.out.use(request_size, now);
 }
 
+std::chrono::milliseconds
+snc_quota_manager::get_quota_balancer_node_period() const {
+    const auto v = _kafka_quota_balancer_node_period();
+    // zero period in config means do not run balancer
+    if (v == std::chrono::milliseconds::zero()) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(
+          ss::steady_clock_type::duration::max() / 2);
+    }
+    return v;
+}
+
+void snc_quota_manager::maybe_arm_balancer_timer() {
+    static constexpr auto min_sleep_time = 2ms; // TBD config?
+
+    if (_balancer_timer.armed()) {
+        return;
+    }
+    ss::steady_clock_type::time_point arm_until
+      = _balancer_timer_last_ran + get_quota_balancer_node_period();
+    const auto now = ss::steady_clock_type::now();
+    const ss::steady_clock_type::time_point closest_arm_until
+      = now + min_sleep_time;
+    if (arm_until < closest_arm_until) {
```

Comment:
> I think this should be refactored to just run on the configured period (its a bounded property that has a static min right?). So we shouldn't have to worry about the case where the loop runs too tight because the min sleep will bound that.

Static min is not the same: intervals are measured at the beginning of the balancer processing so that the time that it takes for a balancer step to run does not influence its frequency. So even with a static minimum it is possible that (run time of the balancer step) > (balancer period) => no delay between balancer runs. This is when this code is supposed to intervene to prevent the balancer from taking too much share of the core0 compute.

Reply:
> Furthermore I don't think its an issue if the algorithm takes longer to run then expected and therefore doesn't run every Xms

I was thinking about high load scenarios. If the reactor is really busy, it can take signinficantly longer for cross shard calls to complete. If we don't try to keep the balancer frequency stable, this may cause the situations like "balancer stops working under high load" (actully it doesn't but only works much slower) -- while the high load scenarios is where it is expected to work in first place.

The next step to make balancer less sensitive to high load is to allow multiple balancer steps to run in parallel, but I have left that improvement for later.

Reply:
there should be no assumptions about _when_ code is going to be executed.

Reply:
> there should be no assumptions about _when_ code is going to be executed

Could you please elaborate?

Reply:
correctness shouldn't be predicated on the timer running at a precise time

Reply:
Yeah i'm still kind of on the fence about this one. To me it makes sense that the only thing we should guarantee, is that invocations of the algorithm occur at least `interval_ms` between each other.

In general (with this fiber/timer design pattern) do we ever care when the time it takes for one iteration of a fiber exceeds its periodicity interval? I don't think its an explicitly handled condition in other places of the codebase today @dotnwat am I right in assuming this?

I think another pattern to consider might be to let the loop run at the configured interval, but maybe have the loop just do no work and return if its observed this condition occur.

Reply:
> correctness shouldn't be predicated on the timer running at a precise time

> To me it makes sense that the only thing we should guarantee, is that invocations of the algorithm occur at least `interval_ms` between each other.

In this case it's not _correctness_, it's _precision_ of the implemented controls. If balancer frequency fluctuates because of the time it takes for an iteration to complete, that does not render the balancer effort incorrent, but the response time/magnitude will be somewhat (slightly or not) different from what may be expected from the current configuration.